### PR TITLE
use contact name in RecipientElement for sending

### LINKED ISSUE #4268: include names in `recipient` field when calling FES (infoDiv)
I don't think the change I've done will work @tomholub
Browser-extension is stripping out the names of the Google contacts - and sending just the email address.
```json
{"associateReplyToken":"e24d308e-6595-42f0-965f-282038d9247d","from":"jb@flowcrypt.com","to":["jerbell@gmx.com"],"cc":[],"bcc":[]}
```
I have `jerbell@gmx.com` as a contact with a name.

FES can deal with the name and email address for all the from/to fields in the `Joe Bloggs <Joe.Bloggs@email.com>`.  Did you want to raise an issue with browser extension?

_Originally posted by @Jerbell in https://github.com/FlowCrypt/flowcrypt-enterprise-on-prem/issues/2153#issuecomment-1018210504_

### PR DESCRIPTION
**Pull request by rrrooommmaaa**
This PR loads names to be added to email into RecipientElement

close #4268 

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by rrrooommmaaa** [comment]
There are 3 places in the code to actually display names in recipient boxes. I can do this in this PR or later.
`AddNameToEmail` is only used for "From" now, Perhaps, we can eliminate it in #4270 ?

**Comment by rrrooommmaaa** [comment]
Hi @tomholub 
Though all tests pass, I'm a bit worried if I missed something. Do we need more tests for this?
I mainly added Names to `RecipientElement`, excluded 'wrong' addresses from evaluation.
Also, `setEmailsPreview` caused deadlock (after my changes somehow) when waiting for evaluation to finish, so now simply abort if evaluation is in progress, and is called again after each evaluation

**Comment by rrrooommmaaa** [comment]
ping @tomholub

**Comment by rrrooommmaaa** [comment]
image

**Comment by rrrooommmaaa** [comment]
Name is now displayed name in "email preview" if available
image

**Comment by tomholub** [comment]
> Hi @tomholub Though all tests pass, I'm a bit worried if I missed something. Do we need more tests for this? I mainly added Names to `RecipientElement`, excluded 'wrong' addresses from evaluation. Also, `setEmailsPreview` caused deadlock (after my changes somehow) when waiting for evaluation to finish, so now simply abort if evaluation is in progress, and is called again after each evaluation

I'll take a look.

As for the screenshots, they look good except the bolded part should have smaller weight. So instead of `weight: bold` could try `weight: 300` or `400` or similar, to be somewhere in between.

File: extension/chrome/elements/compose-modules/compose-recipients-module.ts [hunk 1]
```diff
@@ -677,8 +696,6 @@ export class ComposeRecipientsModule extends ViewModule<ComposeView> {
         contactItems.removeClass('active');
         $(this).addClass('active');
       });
-      this.view.S.cached('contacts').find('ul li.auth_contacts').click(this.view.setHandler(() =>
-        this.authContacts(this.view.acctEmail), this.view.errModule.handle(`authorize contact search`)));
```
Comment (tomholub) [inline extension/chrome/elements/compose-modules/compose-recipients-module.ts:681]: Was this functionality replicated elsewhere? If it wasn't, and tests still pass, then we need to write a test for this (unless this was somehow already duplicate?)

File: extension/chrome/elements/compose-modules/compose-recipients-module.ts [hunk 2]
```diff
@@ -804,95 +827,85 @@ export class ComposeRecipientsModule extends ViewModule<ComposeView> {
...
-      this.canReadEmails = true;
-      this.view.scopes.readContacts = true;
-      this.view.scopes.readOtherContacts = true;
-      await this.searchContacts(this.view.S.cached('input_to'));
-    } else if (authRes.result === 'Denied' || authRes.result === 'Closed') {
-      await Ui.modal.error('FlowCrypt needs this permission to search your contacts on Gmail. Without it, FlowCrypt will keep a separate contact list.');
-    } else {
-      await Ui.modal.error(Lang.general.somethingWentWrongTryAgain(!!this.view.fesUrl));
-    }
-  };
```
Comment (tomholub) [inline extension/chrome/elements/compose-modules/compose-recipients-module.ts:823]: Similar here - I'm surprised to see it removed. See it in action here https://github.com/FlowCrypt/flowcrypt-browser/pull/4121#issuecomment-974508721 (or not?)

File: extension/chrome/elements/compose-modules/compose-recipients-module.ts [hunk 3]
```diff
@@ -804,95 +827,85 @@ export class ComposeRecipientsModule extends ViewModule<ComposeView> {
...
-
+  // todo: I guess we can combine this with reRenderRecipientFor
```
Comment (tomholub) [inline extension/chrome/elements/compose-modules/compose-recipients-module.ts:830]: Leave it to your judgement, unsure

File: extension/chrome/elements/compose-modules/compose-send-btn-module.ts [hunk 4]
```diff
@@ -224,30 +225,24 @@ export class ComposeSendBtnModule extends ViewModule<ComposeView> {
     }
   };
 
-  private addNamesToMsg = async (msg: SendableMsg): Promise<void> => {
+  private addNameToEmail = async (email: string): Promise<string> => {
```
Comment (tomholub) [inline extension/chrome/elements/compose-modules/compose-send-btn-module.ts:228]: Let's call this `formatSenderEmailAsMimeString` (is that what it does?)

File: extension/chrome/elements/compose-modules/formatters/signed-msg-mail-formatter.ts [hunk 5]
```diff
@@ -40,8 +41,8 @@ export class SignedMsgMailFormatter extends BaseMailFormatter {
       // Removing them here will prevent Gmail from screwing up the signature
       newMsg.plaintext = newMsg.plaintext.split('\n').map(l => l.replace(/\s+$/g, '')).join('\n').trim();
       const signedData = await MsgUtil.sign(signingPrv, newMsg.plaintext);
-      const allContacts = [...newMsg.recipients.to || [], ...newMsg.recipients.cc || [], ...newMsg.recipients.bcc || []];
-      ContactStore.update(undefined, allContacts, { lastUse: Date.now() }).catch(Catch.reportErr);
+      const recipients = Value.arr.unique(Object.values(newMsg.recipients).reduce((a, b) => a.concat(b), []).map(x => x.email));
```
Comment (tomholub) [inline extension/chrome/elements/compose-modules/formatters/signed-msg-mail-formatter.ts:44]: Since this line `Value.arr.unique(Object.values(newMsg.recipients).reduce((a, b) => a.concat(b), []).map(x => x.email));` repeats a few times across the files, we could give it a name and make a method to describe what it does. I suppose `getUniqueRecipientEmails` ?

File: extension/css/cryptup.css [hunk 6]
```diff
@@ -1564,6 +1564,15 @@ table#compose td.recipients-inputs > div#input_addresses_container div .recipien
 table#compose td.recipients-inputs > div#input_addresses_container div .recipients > span.failed { border: 1px dashed #a44; }
 table#compose td.recipients-inputs > div#input_addresses_container div .recipients > span.failed .close-icon { opacity: 1; }
 
+.recipient-name {
+  font-weight: bold;
```
Comment (tomholub) [inline extension/css/cryptup.css:1568]: Can try various numeric weights here to find a bit thinner one.

**Review by tomholub — COMMENTED** [review]
Quite a change this ended up being. It's not bad. Comments above

**Comment by rrrooommmaaa** [comment]
> As for the screenshots, they look good except the bolded part should have smaller weight. So instead of `weight: bold` could try `weight: 300` or `400` or similar, to be somewhere in between.

is weight 500 looking good?
image

**Comment by tomholub** [comment]
Better :+1:

File: extension/chrome/elements/compose-modules/compose-recipients-module.ts (hunk 2 above)
Comment (rrrooommmaaa) [reply extension/chrome/elements/compose-modules/compose-recipients-module.ts:823]: if you mean this button "Enable Google Conact Search" -- it is still there.
image
It's defined and handled in another piece of code https://github.com/FlowCrypt/flowcrypt-browser/blob/33b617999bc915a68752d6e85df912bd2d969619/extension/chrome/elements/compose-modules/compose-recipients-module.ts#L726

File: extension/chrome/elements/compose-modules/compose-recipients-module.ts (hunk 3 above)
Comment (rrrooommmaaa) [reply extension/chrome/elements/compose-modules/compose-recipients-module.ts:830]: Better in some other PR

File: extension/chrome/elements/compose-modules/formatters/signed-msg-mail-formatter.ts (hunk 5 above)
Comment (tomholub) [reply extension/chrome/elements/compose-modules/formatters/signed-msg-mail-formatter.ts:44]: You could try `compose-recipients-module.ts` as a public method, then access it in other modules as `this.view.recipientsModule.uniqueRecipientEmails(newMsg.recipients)` or similar.

File: extension/chrome/elements/compose-modules/compose-send-btn-module.ts (hunk 4 above)
Comment (rrrooommmaaa) [reply extension/chrome/elements/compose-modules/compose-send-btn-module.ts]: done

File: extension/chrome/elements/compose-modules/formatters/signed-msg-mail-formatter.ts (hunk 5 above)
Comment (rrrooommmaaa) [reply extension/chrome/elements/compose-modules/formatters/signed-msg-mail-formatter.ts]: Done

Comment (rrrooommmaaa) [reply extension/chrome/elements/compose-modules/formatters/signed-msg-mail-formatter.ts]: Ok, let me fix it...

Comment (rrrooommmaaa) [reply extension/chrome/elements/compose-modules/formatters/signed-msg-mail-formatter.ts]: oh no... looks like recipientsModule isn't available from formatters

**Review by tomholub — APPROVED** [review]
